### DISCUSSION TITLE
How can I add files with persistence?

### DISCUSSION THREAD

**Discussion by shandshellin** [DustinBrett/daedalOS#381] [Q&A] [answered]
Is there any way to add additional emulators for roms?

**Comment by CorruptionVR** [comment]
which emulator do you mean?

**Comment by shandshellin** [comment]
the bundle as in what retroarch houses. There is already dos emulation in daedalos which is fantastic but i would love to run snes, mame, etc. Trying to get that all in web instance. Do game controllers work via the browser? Could i map a directory via docker so i could access me dos games from a share?

**Comment by CorruptionVR** [comment]
> i would love to run snes, mame, etc.

There is emulator.js built into the OS

> Do game controllers work via the browser?

I have not tried that yet, but emulator.js does have gamepad support by itself

> Could i map a directory via docker so i could access me dos games from a share?

I have no idea

**Comment by DustinBrett** [comment]
Indeed [Emulator.js](https://github.com/EmulatorJS/EmulatorJS) already has support for lots of roms (.32x, .a26, .a52, .a78, .gb, .gba, .gbc, .gen, .gg, .j64, .jag, .lnx, .n64, .nds, .nes, .ngc, .ngp, .pce, .sfc, .smc, .smd, .sms, .v64, .vb, .vboy, .ws, .wsc, .z64) built into the OS.

And for mapping of folders, there is the ability to share external folders via the https://developer.mozilla.org/en-US/docs/Web/API/File_System_API, so if the browser has access to folders on the host machine, they can be shared inside daedalOS.

SNES is supported but MAME not because I have excluded it for now, it was slightly more complex to handle rom files.

I have used a game controller once so it can work, but I haven't done much testing and it's all based on whatever Emulator.js supports.

**Comment by CorruptionVR** [comment]
Maybe you could add a feature to connect to one drive or Google Drive in the files app

**Comment by DustinBrett** [comment]
> Maybe you could add a feature to connect to one drive or Google Drive in the files app

Ya that is the plan eventually, to integrate One Drive, and eventually others like Google/Dropbox. I haven't had a lot of personal use for the feature so I have not yet added it, and also I have avoided features which will require forked users to have their own API keys to make them work.

In a perfect world I would not need an API key for these features and I could just have the users give their creds when needed, but because that isn't how it works I would need to have my own API key and be registered with MS/Google.

**Comment by shandshellin** [comment]
you are correct, the roms all work except my mame roms. WIsh i could get unreal tournament on this!

**Comment by DustinBrett** [comment]
> you are correct, the roms all work except my mame roms. WIsh i could get unreal tournament on this!

Ya I will probably add MAME one day, but the roms I had examples of were multi file and for now I decided to focus on roms with single files.

As for UT, I do want to get it added one day, but the versions I have seen online don't work too well and I don't yet know the process to port it myself.

A demo is here of what they had many years ago: https://icculus.org/ut99-emscripten/flyby/wasm/

**Comment by shandshellin** [comment]
That demo is what makes me hopeful. My mame roms are all over the place as well. Any chance of maybe some open office?

**Comment by DustinBrett** [comment]
> That demo is what makes me hopeful. My mame roms are all over the place as well. Any chance of maybe some open office?

Ya Open Office is another possibility, the main reason I haven't done that one yet is the port I have seen is a somewhat large file and also it's not setup well to be interactive with daedalOS.

Here is the demo I have seen although it's LibreOffice instead of Open Office: https://wasm-test.libreoffice.org/

**Comment by shandshellin** [comment]
I hear ya, i've been trying to run my own web based open office but since my cluster is solely arm the options are limited. An all in one web os for the basics is the ideal.

**Comment by shandshellin** [comment]
I tried binding a volume with docker, no error but no folder either, tried a new folder and an existing one. Is the actual folder path different then what is shown in the OS? What is the root folder path?

**Reply by DustinBrett** [reply]
I am not totally clear what you have tried but to clarify what I am saying is possible, I mean doing this:

1. 
![image](https://github.com/DustinBrett/daedalOS/assets/16656318/aa084c72-e984-4b90-91e6-2276d7180b4f)

2. 
![image](https://github.com/DustinBrett/daedalOS/assets/16656318/ebc44cd2-d165-4bbd-9585-6e535bdbd6c6)

3. 
![image](https://github.com/DustinBrett/daedalOS/assets/16656318/705ad88c-b9ad-4b03-89ea-cb68cf9fd50d)

4. 
![image](https://github.com/DustinBrett/daedalOS/assets/16656318/5780f1cb-10ea-4d0c-8ac4-39bb8e91f682)

**Comment by shandshellin** [comment]
I 'm not seeing the "map directory" option. Maybe the guy who built the docker image isn't using the latest build from you.

**Comment by CorruptionVR** [comment]
wait, are you using the dustinbrett.com website?

**Comment by shandshellin** [comment]
no this is hosted on my own cluster, i can bind the drive i need and can see it in the console of the contaioner running daedalos. It does not add my custom folder to the desktop itself, I tried taking over the public/Videos folder and replacing it with my bound folder which again o can browse in the console and i can see the folder is different in the OS but not showing any of my files in that bound folder.

**Reply by DustinBrett** [reply]
The mapping (binding) you are referring to is using File System Access API (https://developer.mozilla.org/en-US/docs/Web/API/File_System_API) which requires a secure context (HTTPS) in the browser. Docker is only running as a web server hosting static files. And I wrote the Dockerfile myself, it's very simple because the code is all client side and all that matters is what the browser can see.

**Comment by shandshellin** [comment]
SO if i copy a file to the desktop in daedalos, is it only stored in the browser cache? If so then when this reboots, it loses any files etc copied top it? That's what happened when i first started playingf with it. This is the docker imnage i am using dcook0820/daedalos as it's the only one that supports arm. No add dir menu in there. I need an option for persistent storage.

**Reply by DustinBrett** [reply]
Ya I don't have direct access to the file system as the files are hosted on a web server only. To give access it is done via the browser. You could open the browser and map a directory which points back into the docker container. You just need to make sure the files are hosted on a HTTPS connection to get secure context.

**Comment by shandshellin** [comment]
If it's pulling static files, what is the root path for the static files? I should be able to bind to one of the static folders and then it would host it out.

**Reply by DustinBrett** [reply]
The folder is the root folder where `index.html` would be, and `public` is the source also when being built. But likely mapping that folder will not work because the file system is created on build into a JSON file.

**Comment by shandshellin** [comment]
does that mean that all the user folders it uses are also in that JSON file? All the fodlers i'm seeing are just for building said JSON file?

**Reply by DustinBrett** [reply]
The JSON file is just the directory structure which is used by BrowserFS. Your best bet is mapping the folder and having that point back to Docker.

**Comment by shandshellin** [comment]
Is there a way to keep the locally coppied data persistent in the browser? Like if i clear my browser cahce for example will any copied files vanish?

**Reply by DustinBrett** [reply]
Only mapped folders would retain data, otherwise clearing the browser would clear all changes in IndexedDb.

**Comment by shandshellin** [comment]
I'm not seeing an option to map from daedalos, the contextr menu in mine does not have "add dir" 

![daedalos](https://github.com/DustinBrett/daedalOS/assets/25377845/edc28175-5e53-4816-922e-b3f30b228280)

**Reply by DustinBrett** [reply]
You need to show me the URL you are using also to clarify if you are on secure context.

**Comment by shandshellin** [comment]
I see the issue, the connection on my lan is not https and therefor won't give me the bind option. Works externally though. I guess i need to sort that local https.

**Comment by shandshellin** [comment]
is there an https port hosted already?

**Reply by DustinBrett** [reply]
Yes in scripts I have `dev:ssl` as another command that could be run. So you could change the Dockerfile to run `yarn dev:ssl` instead of `yarn start`.

Maybe the command could be adjusted to do a prod ssl build, but it's not something I have tried as I host on nginx and have my own ssl cert for my personal domain.

**Comment by shandshellin** [comment]
internally http resolves but https does not. if i try to access from outside https works because it is going through my nginx where it has it's cert.

**Reply by DustinBrett** [reply]
I am unsure for this issue, maybe Docker is not letting port 443 through. But that is what you need to do anyway, get the built files hosted on a web server via https with a ssl cert, then File system access will work in secure context and you can map a drive from your browser back into Docker or wherever.

**Comment by shandshellin** [comment, marked as answer]
added a dns entry to resolve to my external domain for that hostname. now i get https and the add directory option.
